### PR TITLE
Fixed e2e tests for HA master.

### PR DESCRIPTION
**Pull request by jszczepkowski**
Set of fixes that allows HA master e2e tests to pass for removal/addition master replicas.

The summary of changes:
- fixed host name in etcd certs,
- added cluster validation after kube-down,
- fixed the number of master replicas in cluster validation,
- made MULTIZONE=true required for HA master deployments, ensured we correctly handle MULTIZONE=true when user wants to create HA master but not kubelets in multiple zones,
- extended verification of master replicas in HA master e2e tests.

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36999)

### CODE REVIEW

**Comment by fgrzadkowski** [comment]
Can you please explain what you are changing in this PR. It's hard to read it without context.

**Comment by dims** [comment]
@k8s-bot gci gke e2e test this

**Comment by jszczepkowski** [comment]
Here is the summary of changes:
- fixed host name in etcd certs,
- added restarting master kubelet when the last replica remains after kube-down,
- added cluster validation after kube-down,
- fixed the number of master replicas in cluster validation,
- made MULTIZONE=true required for HA master deployments, ensured we correctly handle MULTIZONE=true when user wants to create HA master but not kubelets in multiple zones,
- extended verification of master replicas in HA master e2e tests.

File: cluster/gce/util.sh [hunk 1]
```diff
@@ -1395,6 +1414,10 @@ function kube-down() {
         --region "${REGION}" \
         --quiet \
         "${MASTER_NAME}"
+      # We need to restart master's kubelet as it uses the external IP which destination has just been changed.
```
Comment (fgrzadkowski) [inline cluster/gce/util.sh]: As discussed offline we should verify:
* why exactly is it needed
* why we don't need to do it for all other nodes
* what other options do we have? restarting kubelet is pretty intrusive

File: cluster/gce/util.sh [hunk 2]
```diff
@@ -1786,7 +1837,7 @@ function test-setup() {
   # Detect the project into $PROJECT if it isn't set
   detect-project
 
-  if [[ ${MULTIZONE:-} == "true" ]]; then
+  if [[ ${MULTIZONE:-} == "true" && -n ${E2E_ZONES:-} ]]; then
```
Comment (fgrzadkowski) [inline cluster/gce/util.sh]: Why are you adding it?

File: test/e2e/framework/util.go [hunk 3]
```diff
@@ -4011,6 +4011,33 @@ func WaitForClusterSize(c clientset.Interface, size int, timeout time.Duration)
 	return fmt.Errorf("timeout waiting %v for cluster size to be %d", timeout, size)
 }
 
+// waitForMasters waits until the cluster has the desired number of ready masters in it.
+func WaitForMasters(c clientset.Interface, size int, timeout time.Duration) error {
+	for start := time.Now(); time.Since(start) < timeout; time.Sleep(20 * time.Second) {
+		nodes, err := c.Core().Nodes().List(api.ListOptions{FieldSelector: fields.Set{
+			"spec.unschedulable": "true",
+		}.AsSelector()})
```
Comment (fgrzadkowski) [inline test/e2e/framework/util.go]: This might list some nodes which are not masters, e.g. those which are drained.

File: test/e2e/ha_master.go [hunk 4]
```diff
@@ -56,14 +57,21 @@ func verifyRCs(c clientset.Interface, ns string, names []string) {
 }
 
 func createNewRC(c clientset.Interface, ns string, name string) {
-	_, err := newRCByName(c, ns, name, 1, nil)
+	var err error
+	timeout := 5 * time.Minute
```
Comment (fgrzadkowski) [inline test/e2e/ha_master.go]: Why is it needed?

File: cluster/gce/util.sh (hunk 2 above)
Comment (jszczepkowski) [inline cluster/gce/util.sh]: We need to distinguish two cases: 
1. we want to set-up multimaster cluster with kubelets in a single zone (MULTIZONE=true, E2E_ZONES=false) a
2. we want to set-up cluster with kubelets in many zones (MULTIZONE=ture, E2E_ZONES={list of zones for kubelets})

File: test/e2e/framework/util.go (hunk 3 above)
Comment (jszczepkowski) [inline test/e2e/framework/util.go]: Added additional check for prefix.

File: test/e2e/ha_master.go (hunk 4 above)
Comment (jszczepkowski) [inline test/e2e/ha_master.go]: Not needed anymore, removed.

**Comment by jszczepkowski** [comment]
@fgrzadkowski 
Comments applied, PTAL

File: test/e2e/framework/util.go [hunk 5]
```diff
@@ -4011,6 +4011,39 @@ func WaitForClusterSize(c clientset.Interface, size int, timeout time.Duration)
 	return fmt.Errorf("timeout waiting %v for cluster size to be %d", timeout, size)
 }
 
+// waitForMasters waits until the cluster has the desired number of ready masters in it.
+func WaitForMasters(masterPrefix string, c clientset.Interface, size int, timeout time.Duration) error {
```
Comment (fgrzadkowski) [inline test/e2e/framework/util.go]: How about passing a full regexp for master name and remove checking for unschedulable field?

**Comment by fgrzadkowski** [comment]
Just one last small comment. In general lgtm.

File: test/e2e/framework/util.go (hunk 5 above)
Comment (jszczepkowski) [inline test/e2e/framework/util.go]: Fixed

**Comment by dims** [comment]
@k8s-bot node e2e test this

**Comment by jszczepkowski** [comment]
Part of https://github.com/kubernetes/features/issues/48

**Pull request merged**